The initial building block of Keras is a model, and the simplest model is called sequential. 

A sequential Keras model is a linear pipeline (a stack) of neural networks layers. 

This code fragment defines a single layer with 12 artificial neurons, and it expects 8 input variables (also known as
features):

The Sequential model is a linear stack of layers.

In [3]:
from keras.models import Sequential
from keras.layers import Dense
model = Sequential()
model.add(Dense(12, input_dim=8, kernel_initializer='random_uniform'))

Historically, perceptron was the name given to a model having one single linear layer, and as a consequence, if it
has multiple layers, you would call it multilayer perceptron (MLP).

#### Note :

The net is dense, meaning that each neuron in a layer is connected to all neurons located in the previous layer and to all the neurons in the following layer.

# A real example — recognizing handwritten digits

A database of handwritten digits made up of a training set of 60,000 examples and a test set of 10,000 examples.

Each MNIST image is in gray scale, and it consists of 28 x 28 pixels

In [4]:
from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD
from keras.utils import np_utils
np.random.seed(1671) # for reproducibility
# network and training
NB_EPOCH = 200
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # SGD optimizer, explained later in this chapter
N_HIDDEN = 128
VALIDATION_SPLIT=0.2 # how much TRAIN is reserved for VALIDATION
# data: shuffled and split between train and test sets
# 
(X_train, y_train), (X_test, y_test) = mnist.load_data()
#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784
RESHAPED = 784
# 
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# normalize[0,1]
#Typically, the values associated with each pixel are normalized in the range [0, 1] (which means that
#the intensity of each pixel is divided by 255, the maximum intensity value). The output is 10 classes,
#one for each digit.
X_train /= 255
X_test /= 255
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')
# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)

11493376/11490434 [==============================] - 30s 3us/step
60000 train samples
10000 test samples


The final layer is a single neuron with activation function softmax, which is a generalization of the
sigmoid function. Softmax squashes a k-dimensional vector of arbitrary real values into a kdimensional
vector of real values in the range (0, 1). In our case, it aggregates 10 answers provided
by the previous layer with 10 neurons:

In [5]:
# 10 outputs
# final stage is softmax
model = Sequential()
model.add(Dense(NB_CLASSES, input_shape=(RESHAPED,)))
model.add(Activation('softmax'))
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_2 (Dense)              (None, 10)                7850      
_________________________________________________________________
activation_1 (Activation)    (None, 10)                0         
Total params: 7,850
Trainable params: 7,850
Non-trainable params: 0
_________________________________________________________________


In [6]:
model.compile(loss='categorical_crossentropy', optimizer=OPTIMIZER, metrics=['accuracy'])

Once the model is compiled, it can be then trained with the fit() function, which specifies a few
parameters:

    epochs: This is the number of times the model is exposed to the training set. At each iteration, the optimizer tries to adjust the weights so that the objective function is minimized.

    batch_size: This is the number of training instances observed before the optimizer performs a weight update.

#### Training a model in Keras

In [7]:
history = model.fit(X_train, Y_train,batch_size=BATCH_SIZE, epochs=NB_EPOCH,verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

Train on 48000 samples, validate on 12000 samples
Epoch 1/200
48000/48000 [==============================] - 3s 68us/step - loss: 1.3633 - acc: 0.6796 - val_loss: 0.8904 - val_acc: 0.8246
Epoch 2/200
48000/48000 [==============================] - 2s 41us/step - loss: 0.7913 - acc: 0.8272 - val_loss: 0.6572 - val_acc: 0.8546
Epoch 3/200
48000/48000 [==============================] - 2s 33us/step - loss: 0.6436 - acc: 0.8497 - val_loss: 0.5625 - val_acc: 0.8681
Epoch 4/200
48000/48000 [==============================] - 2s 41us/step - loss: 0.5717 - acc: 0.8602 - val_loss: 0.5098 - val_acc: 0.8765
Epoch 5/200
48000/48000 [==============================] - 2s 41us/step - loss: 0.5276 - acc: 0.8678 - val_loss: 0.4758 - val_acc: 0.8826
Epoch 6/200
48000/48000 [==============================] - 2s 37us/step - loss: 0.4973 - acc: 0.8726 - val_loss: 0.4515 - val_acc: 0.8866
Epoch 7/200
48000/48000 [==============================] - 2s 47us/step - loss: 0.4748 - acc: 0.8775 - val_loss: 0.4333 - 

48000/48000 [==============================] - 2s 36us/step - loss: 0.3143 - acc: 0.9128 - val_loss: 0.3022 - val_acc: 0.9151
Epoch 60/200
48000/48000 [==============================] - 2s 34us/step - loss: 0.3136 - acc: 0.9129 - val_loss: 0.3016 - val_acc: 0.9161
Epoch 61/200
48000/48000 [==============================] - 2s 34us/step - loss: 0.3130 - acc: 0.9133 - val_loss: 0.3011 - val_acc: 0.9158
Epoch 62/200
48000/48000 [==============================] - 2s 36us/step - loss: 0.3123 - acc: 0.9131 - val_loss: 0.3007 - val_acc: 0.9151
Epoch 63/200
48000/48000 [==============================] - 2s 31us/step - loss: 0.3117 - acc: 0.9136 - val_loss: 0.3003 - val_acc: 0.9156
Epoch 64/200
48000/48000 [==============================] - 2s 32us/step - loss: 0.3110 - acc: 0.9137 - val_loss: 0.2997 - val_acc: 0.9158
Epoch 65/200
48000/48000 [==============================] - 2s 33us/step - loss: 0.3105 - acc: 0.9137 - val_loss: 0.2992 - val_acc: 0.9159
Epoch 66/200
48000/48000 [==============

Epoch 118/200
48000/48000 [==============================] - 2s 32us/step - loss: 0.2902 - acc: 0.9192 - val_loss: 0.2846 - val_acc: 0.9212
Epoch 119/200
48000/48000 [==============================] - 2s 34us/step - loss: 0.2900 - acc: 0.9191 - val_loss: 0.2844 - val_acc: 0.9212
Epoch 120/200
48000/48000 [==============================] - 2s 33us/step - loss: 0.2898 - acc: 0.9192 - val_loss: 0.2842 - val_acc: 0.9212
Epoch 121/200
48000/48000 [==============================] - 2s 35us/step - loss: 0.2895 - acc: 0.9191 - val_loss: 0.2841 - val_acc: 0.9212
Epoch 122/200
48000/48000 [==============================] - 2s 38us/step - loss: 0.2892 - acc: 0.9192 - val_loss: 0.2840 - val_acc: 0.9212
Epoch 123/200
48000/48000 [==============================] - 2s 35us/step - loss: 0.2890 - acc: 0.9194 - val_loss: 0.2838 - val_acc: 0.9211
Epoch 124/200
48000/48000 [==============================] - 2s 35us/step - loss: 0.2888 - acc: 0.9197 - val_loss: 0.2837 - val_acc: 0.9210
Epoch 125/200
48000/

48000/48000 [==============================] - 2s 37us/step - loss: 0.2793 - acc: 0.9220 - val_loss: 0.2775 - val_acc: 0.9232
Epoch 177/200
48000/48000 [==============================] - 2s 41us/step - loss: 0.2791 - acc: 0.9219 - val_loss: 0.2774 - val_acc: 0.9234
Epoch 178/200
48000/48000 [==============================] - 2s 35us/step - loss: 0.2790 - acc: 0.9221 - val_loss: 0.2774 - val_acc: 0.9228
Epoch 179/200
48000/48000 [==============================] - 2s 32us/step - loss: 0.2788 - acc: 0.9221 - val_loss: 0.2773 - val_acc: 0.9232
Epoch 180/200
48000/48000 [==============================] - 2s 39us/step - loss: 0.2787 - acc: 0.9221 - val_loss: 0.2771 - val_acc: 0.9235
Epoch 181/200
48000/48000 [==============================] - 2s 35us/step - loss: 0.2785 - acc: 0.9223 - val_loss: 0.2770 - val_acc: 0.9232
Epoch 182/200
48000/48000 [==============================] - 2s 38us/step - loss: 0.2784 - acc: 0.9220 - val_loss: 0.2769 - val_acc: 0.9231
Epoch 183/200
48000/48000 [=======

#### Note:
We reserved part of the training set for validation. The key idea is that we reserve a
part of the training data for measuring the performance on the validation while
training. This is a good practice to follow for any machine learning task, which we
will adopt in all our examples.
Once the model is trained, we can evaluate it

In [8]:
score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test score:", score[0])
print('Test accuracy:', score[1])

10000/10000 [==============================] - 1s 51us/step
Test score: 0.27738585189282894
Test accuracy: 0.9227


In [9]:
score

[0.27738585189282894, 0.9227]

A first improvement is to add additional layers to our network. 

So, after the input layer, we have a first dense layer with the N_HIDDEN neurons and an activation function relu. This additional layer is
considered hidden because it is not directly connected to either the input or the output. 

After the first hidden layer, we have a second hidden layer, again with the N_HIDDEN neurons, followed by an output
layer with 10 neurons, each of which will fire when the relative digit is recognized.

In [10]:
# M_HIDDEN hidden layers
# 10 outputs
# final stage is softmax
model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,)))
model.add(Activation('relu'))
model.add(Dense(N_HIDDEN))
model.add(Activation('relu'))
model.add(Dense(NB_CLASSES))
model.add(Activation('softmax'))
model.summary()
model.compile(loss='categorical_crossentropy',optimizer=OPTIMIZER,metrics=['accuracy'])

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 128)               100480    
_________________________________________________________________
activation_2 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_4 (Dense)              (None, 128)               16512     
_________________________________________________________________
activation_3 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_5 (Dense)              (None, 10)                1290      
_________________________________________________________________
activation_4 (Activation)    (None, 10)                0         
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________________________________________________________________


In [11]:
#Train the model
history = model.fit(X_train, Y_train,batch_size=BATCH_SIZE, epochs=NB_EPOCH,verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

Train on 48000 samples, validate on 12000 samples
Epoch 1/200
48000/48000 [==============================] - 4s 91us/step - loss: 1.5205 - acc: 0.6090 - val_loss: 0.7762 - val_acc: 0.8380
Epoch 2/200
48000/48000 [==============================] - 3s 64us/step - loss: 0.6140 - acc: 0.8474 - val_loss: 0.4608 - val_acc: 0.8807
Epoch 3/200
48000/48000 [==============================] - 3s 64us/step - loss: 0.4453 - acc: 0.8795 - val_loss: 0.3783 - val_acc: 0.8962
Epoch 4/200
48000/48000 [==============================] - 3s 70us/step - loss: 0.3835 - acc: 0.8925 - val_loss: 0.3394 - val_acc: 0.9055
Epoch 5/200
48000/48000 [==============================] - 3s 67us/step - loss: 0.3489 - acc: 0.9009 - val_loss: 0.3155 - val_acc: 0.9101
Epoch 6/200
48000/48000 [==============================] - 3s 70us/step - loss: 0.3247 - acc: 0.9076 - val_loss: 0.2967 - val_acc: 0.9161
Epoch 7/200
48000/48000 [==============================] - 3s 71us/step - loss: 0.3066 - acc: 0.9127 - val_loss: 0.2816 - 

48000/48000 [==============================] - 3s 62us/step - loss: 0.0872 - acc: 0.9761 - val_loss: 0.1135 - val_acc: 0.9672
Epoch 60/200
48000/48000 [==============================] - 3s 70us/step - loss: 0.0859 - acc: 0.9761 - val_loss: 0.1124 - val_acc: 0.9673
Epoch 61/200
48000/48000 [==============================] - 3s 70us/step - loss: 0.0846 - acc: 0.9766 - val_loss: 0.1124 - val_acc: 0.9673
Epoch 62/200
48000/48000 [==============================] - 3s 63us/step - loss: 0.0833 - acc: 0.9769 - val_loss: 0.1111 - val_acc: 0.9677
Epoch 63/200
48000/48000 [==============================] - 3s 73us/step - loss: 0.0820 - acc: 0.9772 - val_loss: 0.1107 - val_acc: 0.9679
Epoch 64/200
48000/48000 [==============================] - 3s 72us/step - loss: 0.0807 - acc: 0.9780 - val_loss: 0.1099 - val_acc: 0.9682
Epoch 65/200
48000/48000 [==============================] - 3s 68us/step - loss: 0.0794 - acc: 0.9781 - val_loss: 0.1085 - val_acc: 0.9679
Epoch 66/200
48000/48000 [==============

Epoch 118/200
48000/48000 [==============================] - 3s 64us/step - loss: 0.0391 - acc: 0.9905 - val_loss: 0.0906 - val_acc: 0.9743
Epoch 119/200
48000/48000 [==============================] - 3s 62us/step - loss: 0.0386 - acc: 0.9906 - val_loss: 0.0908 - val_acc: 0.9745
Epoch 120/200
48000/48000 [==============================] - 4s 73us/step - loss: 0.0382 - acc: 0.9909 - val_loss: 0.0909 - val_acc: 0.9739
Epoch 121/200
48000/48000 [==============================] - 3s 61us/step - loss: 0.0376 - acc: 0.9910 - val_loss: 0.0912 - val_acc: 0.9729
Epoch 122/200
48000/48000 [==============================] - 3s 68us/step - loss: 0.0372 - acc: 0.9911 - val_loss: 0.0906 - val_acc: 0.9742
Epoch 123/200
48000/48000 [==============================] - 4s 81us/step - loss: 0.0368 - acc: 0.9913 - val_loss: 0.0899 - val_acc: 0.9746
Epoch 124/200
48000/48000 [==============================] - 3s 64us/step - loss: 0.0363 - acc: 0.9918 - val_loss: 0.0900 - val_acc: 0.9743
Epoch 125/200
48000/

48000/48000 [==============================] - 3s 65us/step - loss: 0.0199 - acc: 0.9968 - val_loss: 0.0891 - val_acc: 0.9747
Epoch 177/200
48000/48000 [==============================] - 3s 65us/step - loss: 0.0196 - acc: 0.9969 - val_loss: 0.0898 - val_acc: 0.9747
Epoch 178/200
48000/48000 [==============================] - 3s 63us/step - loss: 0.0194 - acc: 0.9970 - val_loss: 0.0892 - val_acc: 0.9743
Epoch 179/200
48000/48000 [==============================] - 3s 72us/step - loss: 0.0192 - acc: 0.9969 - val_loss: 0.0895 - val_acc: 0.9743
Epoch 180/200
48000/48000 [==============================] - 3s 69us/step - loss: 0.0191 - acc: 0.9969 - val_loss: 0.0895 - val_acc: 0.9745
Epoch 181/200
48000/48000 [==============================] - 3s 68us/step - loss: 0.0188 - acc: 0.9971 - val_loss: 0.0894 - val_acc: 0.9744
Epoch 182/200
48000/48000 [==============================] - 3s 66us/step - loss: 0.0186 - acc: 0.9972 - val_loss: 0.0894 - val_acc: 0.9743
Epoch 183/200
48000/48000 [=======

In [12]:
score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test score:", score[0])
print('Test accuracy:', score[1])

10000/10000 [==============================] - 1s 82us/step
Test score: 0.08070503973413724
Test accuracy: 0.9766


#### Note:

By adding two hidden layers, we reached 99.80% accuracy on the training set, 97.43% accuracy on validation, and 97.66% accuracy on the test.

# Further improving the simple net in Keras with dropout

We decide to randomly drop with the dropout probability some
of the values propagated inside our internal dense network of hidden layers. In machine learning, this
is a well-known form of regularization. Surprisingly enough, this idea of randomly dropping a few
values can improve our performance: